[WIP] Sharing multiple experiments

karajan1001 opened the pull request:
fix #5489 
1. exp push/pull: accept multiple exp as input
2. exp push/pull: can accept the same arguments as the `exp ls` and work on exactly the same exps as shown in `ls`
3. add a new flag branch for branch name
4. add new argument max_count for history of a rev

wait for https://github.com/iterative/scmrepo/pull/13


* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [ ] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

pmrowla commented on dvc/repo/experiments/utils.py:
```diff
@@ -203,3 +203,59 @@ def check_ref_format(scm: "Git", ref: ExpRefInfo):
         )
+
+
+def name2exp_ref(
```
If you look at the implementation for `resolve_exp_ref`, it will end up iterating over all local or remote names (when we use functions like `exp_refs_by_name` around like 165). So this `name2exp_ref` implementation will make us do that full list fetch/iteration for each name in `exp_names`.

I think this whole workflow needs to be refactored so that we only have to iterate over the full list of local or remote experiments once:

- the `(remote)_exp_refs_by_name` methods should accept multiple names, and return a dict mapping a name we searched for to the list of matches for that name.
- `resolve_exp_ref` would also accept multiple names, and raise an exception as soon as one of them could not be resolved (i.e. as soon as we get a dict item with `len(values) != 1`)

pmrowla reviewed (changes requested):
So after reviewing this, I think it's clear that the scope of #5489 is actually larger than expected, given that we (ideally) want consistent flags and behavior for all of the relevant commands: `show`, `list`, `push`, `pull` (and potentially even `remove`?)

I think it may be best to split this task up into separate PR's:

1. add `--rev` and updated `-n=-1` behavior  to `exp show`
    * This PR would implement the "get baseline revs per new command flags" utils methods
2. add the new flags to `exp list`
    * This PR would implement the utils methods for listing local and remote exps from the baseline revs returned in the function added in (1)
3. add the new flags to `push/pull`
    * This PR would use the functionality added in (1) and (2) and implement support for pushing/pulling multiple exps

Potentially considering how to use this functionality in `remove` could then be a follow up issue/discussion.

@dberenbaum @karajan1001 WDYT

karajan1001 commented on dvc/command/experiments.py:
```diff
@@ -881,6 +907,36 @@ def run(self):
 
+parent_list_parser = argparse.ArgumentParser(add_help=False)
+filter_group = parent_list_parser.add_mutually_exclusive_group()
+filter_group.add_argument(
+    "--branch",
+    type=str,
+    default=None,
+    help=("experiments derived from the specified branch."),
+    metavar="<branch>",
+)
+filter_group.add_argument(
+    "--rev",
+    type=str,
+    default=None,
+    help=("experiments derived from the specified revision."),
+    metavar="<rev>",
+)
```
`resolve_rev` can accept `branch` name as well as `revision` id. So even I didn't change anything, we can still put `branch` name to the argument `--rev` but the problem is that we might need to clarify this to the users.

karajan1001 replied:
Besides this, there is another thing needs to be noticed that, in the new version `dvc exp pull/push` will inherit from `dvc exp ls` a new behavior that can accept no argument and push all exp based on `HEAD` .

pmrowla replied:
`--rev` is already used in other DVC commands for providing branch and tag names (in addition to actual SHAs), so I don't think users will be confused by that.

Regarding `push/pull` accepting no args and pushing everything based on `HEAD` as the default behavior, I'm not sure that is desired here (pushing and pulling multiple exps should be explicit). In this case, I think the correct behavior would be to require `push/pull -w/--workspace` to explicitly state that you want to push all exps based on your current workspace's HEAD.

karajan1001 commented on dvc/command/experiments.py:
```diff
@@ -881,6 +907,36 @@ def run(self):
         return 0
 
 
+parent_list_parser = argparse.ArgumentParser(add_help=False)
```
BTW, too many classes and cmds in `command.experiments` can we separate them into different files?

pmrowla replied:
Yes that's fine

karajan1001 commented:
> So after reviewing this, I think it's clear that the scope of #5489 is a bit larger than expected, given that we (ideally) want consistent flags and behavior for all of the relevant commands: show, list, push, pull (and potentially even remove?)
>
> I think it may be best to split this task up into separate PR's:
>
> add --rev and updated -n=-1 behavior to exp show
> This PR would implement the "get baseline revs per new command flags" utils methods (and move/refactor the existing  exp show -n functionality into experiments.utils instead of being specific to experiments.show
> This would allow users to do exp show for specific revs/branches (and ranges via -n) without having to git checkout them first
> Related: #5710
> add the new flags (including -A/--all-commits, -a/--all-branches, --all-tags) to exp list
> This PR would implement the utils methods for listing local and remote exps from the baseline revs returned in the function added in (1)
> This PR would also include the resolve_exp_ref refactor to support resolving multiple names at a time
> add the new flags to push/pull
> This PR would use the functionality added in (1) and (2) and implement support for pushing/pulling multiple exps
Potentially considering how to use this functionality in remove could then be a follow up issue/discussion.

@dberenbaum @karajan1001 WDYT

1. Agreed most part. 
2. Yes, I planned to make support for `remove` in the following work. 

Let me first split the original feature request and split this big PR into pieces.

karajan1001 commented:
[![asciicast](https://asciinema.org/a/aoGpfCY5wZnUuOAPwVbTQKPBS.svg)](https://asciinema.org/a/aoGpfCY5wZnUuOAPwVbTQKPBS)
